intercom-cpp/src/detail/utility.h

```diff
@@ -0,0 +1,77 @@
+
+#ifndef INTERCOM_CPP_DETAIL_UTILITY_H
+#define INTERCOM_CPP_DETAIL_UTILITY_H
+
+
+#include <cstring>
+#include <iomanip>
+#include <sstream>
+#include <type_traits>
+
+namespace intercom
+{
+namespace detail
+{
+
+    /**
+     * @brief Tests whether the machine is little endian or not.
+     *
+     * @return true The machine is little endian.
+     * @return false The machine is big endian.
+     */
+    inline bool is_little_endian()
+    {
+        // Create a copy of the data with memcpy.
+        // Directly casting with e.g. reinterpret_cast will lead to UB.
+        int test_value = 1;
+        std::array< uint8_t, sizeof( int ) > asBytes;
+        std::memcpy( &asBytes, &test_value, sizeof( int ) );
+        return asBytes[ 0 ] == 1;
+    }
+
+    /**
+     * @brief Writes the specified binary data into the stream as hex.
+     *
+     * @param stream Target stream.
+     * @param converter The binary data to write.
+     * @return std::ostream& Returns the stream.
+     */
+    template< typename TData >
+    inline std::ostream& write_as_hex(
+        std::ostream& stream,
+        TData data
+    )
+    {
+        // Create a copy of the data with memcpy.
+        // Directly casting with e.g. reinterpret_cast will lead to UB.
+        std::array< uint8_t, sizeof(TData) > asBytes;
+        std::memcpy( &asBytes, &data, sizeof( TData ) );
+
+        // Determine the order in which we need to print the data.
+        stream << std::hex << std::uppercase;
+        if( is_little_endian() )
+        {
+            // On little-endian machine the bytes are on reverse order.
+            for( auto byte = asBytes.rbegin(); byte != asBytes.rend(); ++byte )
+            {
+                // Explicit cast required, otherwise byte is treated as unsigned char.
+                stream << static_cast< unsigned int >( ( *byte & 0xF0 ) >> 4 );
+                stream << static_cast< unsigned int >( *byte & 0x0F );
+            }
```

Comment:
Bytes of _individual fields_ are on reverse order. The fields themselves are in the same order. So if `TData` is a struct, this doesn't apply.

Namely with `struct S { int a; int b }`, `&s.a` should always be the start of the struct, with `&s.b > &s.a`.

In the use case (like in the PR) where `TData` is an integral type, this isn't a problem. However if this appears in the public API of the C++ library, we might want to hide the template implementation behind an anonymous namespace and add overloaded `write_as_hex` implementations for the compatible (ie. integral) types.

Anyway that can be done at a later point. I think there's also some similar API work to be done in Rust-side.

Reply:
"write_as_hex" is a "detail" of intercom library which is supposed to be private by convention. I want to use it from different places within intercom so I really cannot put it inside anonymous namespace.

Other than unsigned integral types can be excluded with a bit appropriate [template](http://www.cplusplus.com/reference/type_traits/is_unsigned/). I'll implement that soonish.